### PR TITLE
Feat/implement post picture endpoints #BLOG-55

### PR DESCRIPTION
**Pull request by fenohasinalala**

### CODE REVIEW

File: src/main/java/com/blogify/blogapi/constant/FileConstant.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.blogify.blogapi.constant;
+
+import java.time.Duration;
+
+public class FileConstant {
+  public static final Duration URL_DURATION = Duration.ofMinutes(2);
```
Comment (nvagno) [inline src/main/java/com/blogify/blogapi/constant/FileConstant.java:6]: You can put it in FileUserService or do we need a lot of file constant?

File: src/main/java/com/blogify/blogapi/endpoint/security/SecurityConf.java [hunk 2]
```diff
@@ -133,6 +135,14 @@ protected void configure(HttpSecurity http) throws Exception {
         .permitAll()
         .antMatchers(HttpMethod.GET, "/posts/*/comments")
         .permitAll()
+        .antMatchers(HttpMethod.GET, "/posts/*/pictures")
```
Comment (nvagno) [inline src/main/java/com/blogify/blogapi/endpoint/security/SecurityConf.java]: Use a static import

File: src/main/java/com/blogify/blogapi/service/PostFileService.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import org.springframework.stereotype.Service;
+import org.springframework.web.multipart.MultipartFile;
+
+@Service
+@AllArgsConstructor
+public class PostFileService {
+  private final PostService postService;
+  private final PostPictureRepository repository;
+  private final S3Service s3Service;
+  private final BucketComponent bucketComponent;
```
Comment (nvagno) [inline src/main/java/com/blogify/blogapi/service/PostFileService.java]: S3Service and BucketComponent are linked so you should call it in S3Service for more abstraction

File: src/main/java/com/blogify/blogapi/service/PostFileService.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  private PostPicture getPictureWithBucketKey(
+      com.blogify.blogapi.repository.model.PostPicture postPicture) {
+    PostPicture picture = new PostPicture();
+    picture.setId(postPicture.getId());
+    picture.setPostId(postPicture.getPost().getId());
+    picture.setUrl(
+        String.valueOf(
+            bucketComponent.presign(postPicture.getBucketKey(), FileConstant.URL_DURATION)));
```
Comment (nvagno) [inline src/main/java/com/blogify/blogapi/service/PostFileService.java]: should be `s3service.generatePresignedUrl(params)` for example

File: src/main/java/com/blogify/blogapi/endpoint/rest/controller/PostFileController.java [hunk 5]
```diff
@@ -28,9 +29,15 @@ public PostPicture uploadPostPicture(
     return service.uploadPicture(pid, picId, file);
   }
 
+  @DeleteMapping(value = "/posts/{pid}/pictures/{picId}")
+  public PostPicture deletePostPictureById(@PathVariable String pid, @PathVariable String picId
```
Comment (nvagno) [inline src/main/java/com/blogify/blogapi/endpoint/rest/controller/PostFileController.java]: Always the path variable name like `PathVariable("pid")` String pid

File: src/main/java/com/blogify/blogapi/endpoint/rest/controller/PostFileController.java [hunk 6]
```diff
@@ -28,9 +29,15 @@ public PostPicture uploadPostPicture(
...
+      ) {
+    return service.deletePictureById(pid, picId);
+  }
+
   @GetMapping(value = "/posts/{pid}/pictures/{picId}")
   public PostPicture getPostPictureById(@PathVariable String pid, @PathVariable String picId
```
Comment (nvagno) [inline src/main/java/com/blogify/blogapi/endpoint/rest/controller/PostFileController.java]: same here

File: src/main/java/com/blogify/blogapi/file/S3Service.java [hunk 7]
```diff
@@ -32,22 +35,38 @@ public String uploadObjectToS3Bucket(String key, byte[] file) {
     }
   }
 
+  public String deleteS3Object(String key) {
+    DeleteObjectRequest deleteObjectsRequest = DeleteObjectRequest.builder().bucket(bucketName).key(key).build();
+    try {
+      s3Client.deleteObject(deleteObjectsRequest);
+      return key;
+    } catch (AwsServiceException | SdkClientException e) {
```
Comment (nvagno) [inline src/main/java/com/blogify/blogapi/file/S3Service.java:49]: No. Throw exception

File: src/main/java/com/blogify/blogapi/file/S3Service.java [hunk 8]
```diff
@@ -32,22 +35,38 @@ public String uploadObjectToS3Bucket(String key, byte[] file) {
...
-      if (e instanceof AwsServiceException awsException) {
-        statusCode = awsException.statusCode();
-      }
-      if (400 <= statusCode && statusCode < 500) {
-        throw new ApiException(ApiException.ExceptionType.CLIENT_EXCEPTION, e.getMessage());
-      } else {
-        throw new ApiException(ApiException.ExceptionType.SERVER_EXCEPTION, e.getMessage());
-      }
+      s3ErrorHandler(e);
+      return new byte[0];
```
Comment (nvagno) [inline src/main/java/com/blogify/blogapi/file/S3Service.java:62]: same here

File: src/main/java/com/blogify/blogapi/file/S3Service.java [hunk 9]
```diff
@@ -32,22 +35,38 @@ public String uploadObjectToS3Bucket(String key, byte[] file) {
...
+    }
+  }
+
+  private void s3ErrorHandler(Exception e){
```
Comment (nvagno) [inline src/main/java/com/blogify/blogapi/file/S3Service.java]: If you want to use this method, try `Thowable` as returned value, hence you can throw it

File: src/main/java/com/blogify/blogapi/constant/FileConstant.java (hunk 1 above)
Comment (fenohasinalala) [reply src/main/java/com/blogify/blogapi/constant/FileConstant.java:6]: It's used by UserFileService and PostFileService

File: src/main/java/com/blogify/blogapi/service/PostFileService.java (hunk 3 above)
Comment (fenohasinalala) [reply src/main/java/com/blogify/blogapi/service/PostFileService.java]: BucketComponent and S3Service use two different SDKs:
- BucketComponent uses s3-transfer-manager SDK: it only manages real File and Directory from the client side (poja generated, and limited), it's main purpose is to check s3 health.
- S3Service uses s3 SDK: use to upload byte[] and delete files in s3 bucket.

File: src/main/java/com/blogify/blogapi/service/PostFileService.java (hunk 4 above)
Comment (fenohasinalala) [reply src/main/java/com/blogify/blogapi/service/PostFileService.java]: it's poja generated method : bucketComponent.presign()

**Review by nvagno — APPROVED** [review]

**Pull request merged**